synapse/handlers/user_directory.py

```diff
@@ -290,13 +304,13 @@ async def _handle_room_publicity_change(
             room_id
         )
 
-        logger.debug("Change: %r, is_public: %r", change, is_public)
+        logger.debug("Change: %r, is_public: %r", publicity, is_public)
 
-        if change and not is_public:
+        if publicity is MatchChange.now_true and not is_public:
             # If we became world readable but room isn't currently public then
             # we ignore the change
             return
-        elif not change and is_public:
+        elif publicity is MatchChange.now_false and is_public:
             # If we stopped being world readable but are still public,
             # ignore the change
             return
```

Comment:
I think I've spotted another problem.

If a room becomes public, within `_handle_room_publicity_change` we'll fetch `other_users_in_room_with_profiles` by calling `get_users_in_room_with_profiles`. This will read from `room_memberships`, which will contain any per-room nicknames and avatars. We'll then shove those back into the directory by calling `_handle_new_user`.

(The relevant bit of code is just below here, but I couldn't add a comment because it's not part of the diff. GRR)

Reply:
(The above is okay from a privacy point of view, but breaks the desired invariant "the directory only contains public profile information)

Reply:
> If a room becomes public, within _handle_room_publicity_change we'll fetch other_users_in_room_with_profiles by calling get_users_in_room_with_profiles. This will read from room_memberships, which will contain any per-room nicknames and avatars. We'll then shove those back into the directory by calling _handle_new_user.

Yup, so I think `_handle_new_user` is basically doing too much (it shouldn't be involved with display names or avatars).

A room becoming public should have no effect upon the `user_directory` table anymore (all the users will still/already be in there!).